### PR TITLE
Add `ExecutionPathTracer` to repo.

### PR DESCRIPTION
**Pull request by theRoughCode**
This PR addresses the first out of 3 components (Simulator, JavaScript renderer, magic command) required to implement #158 .

This PR adds the `ExecutionPathTracer` files and the extension method to attach to any simulator and trace out the operations performed in a given execution path.

This will be merged in to the feature branch (`feature/path_visualizer`).

Note: This should be merged in after updating the `qsharp-runtime` version to include [this PR](https://github.com/microsoft/qsharp-runtime/pull/301).

### CODE REVIEW

File: src/Core/ExecutionPathTracer/ExecutionPathTracer.cs [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+                .OrderBy(k => k)
+                .Select(k =>
+                {
+                    var qubitDecl = new QubitDeclaration(k);
+                    if (this.classicalRegisters.ContainsKey(k))
+                    {
+                        qubitDecl.NumChildren = this.classicalRegisters[k].Count;
+                    }
+
+                    return qubitDecl;
```
Comment (cgranade) [inline src/Core/ExecutionPathTracer/ExecutionPathTracer.cs:56]: Won't help directly in this case, but check out [C# 9 with-expressions](https://devblogs.microsoft.com/dotnet/welcome-to-c-9-0/#with-expressions).

File: src/Core/ExecutionPathTracer/ExecutionPathTracer.cs [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+        {
+            var t = operation.GetType();
+            if (t == null) return false;
+
+            return t.IsGenericType && t.GetGenericTypeDefinition() == typeof(OperationPartial<,,>);
+        }
+
+        /// <summary>
+        /// Retrieves the <c>QubitRegister</c> associated with the given <c>Qubit</c> or create a new
+        /// one if it doesn't exist.
```
Comment (cgranade) [inline src/Core/ExecutionPathTracer/ExecutionPathTracer.cs:104]: It's a bit confusing for a "Get★" method to have a side-effect of creating something...

File: src/Core/ExecutionPathTracer/ExecutionPathTracer.cs [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+                this.qubitRegisters[qubit.Id] = new QubitRegister(qubit.Id);
+            }
+
+            return this.qubitRegisters[qubit.Id];
+        }
+
+        private List<QubitRegister> GetQubitRegisters(IEnumerable<Qubit> qubits)
+        {
+            return qubits.Select(this.GetQubitRegister).ToList();
+        }
```
Comment (cgranade) [inline src/Core/ExecutionPathTracer/ExecutionPathTracer.cs:119]:
```suggestion
        private List<QubitRegister> GetQubitRegisters(IEnumerable<Qubit> qubits) =>
            qubits.Select(this.GetQubitRegister).ToList();
```

File: src/Core/ExecutionPathTracer/ExecutionPathTracer.cs [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+
+        /// <summary>
+        /// Creates a new <c>ClassicalRegister</c> and associate it with the given <c>Qubit</c>.
+        /// </summary>
+        private ClassicalRegister CreateClassicalRegister(Qubit measureQubit)
```
Comment (cgranade) [inline src/Core/ExecutionPathTracer/ExecutionPathTracer.cs:124]: This feels similar enough to the previous that you really want an extension method on `IDictionary<TKey, TValue>` that gets or creates a key.

File: src/Core/ExecutionPathTracer/ExecutionPathTracer.cs [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+        {
+            var qId = measureQubit.Id;
+
+            if (!this.classicalRegisters.ContainsKey(qId))
+            {
+                this.classicalRegisters[qId] = new List<ClassicalRegister>();
+            }
+
+            var cId = this.classicalRegisters[qId].Count;
+            ClassicalRegister register = new ClassicalRegister(qId, cId);
```
Comment (cgranade) [inline src/Core/ExecutionPathTracer/ExecutionPathTracer.cs:134]:
```suggestion
            var register = new ClassicalRegister(qId, cId);
```

File: src/Core/ExecutionPathTracer/ExecutionPathTracer.cs [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+            return t.IsGenericType && t.GetGenericTypeDefinition() == typeof(OperationPartial<,,>);
+        }
+
+        /// <summary>
+        /// Retrieves the <c>QubitRegister</c> associated with the given <c>Qubit</c> or create a new
+        /// one if it doesn't exist.
+        /// </summary>
+        private QubitRegister GetQubitRegister(Qubit qubit)
+        {
+            if (!this.qubitRegisters.ContainsKey(qubit.Id))
```
Comment (cgranade) [inline src/Core/ExecutionPathTracer/ExecutionPathTracer.cs:108]: A lot of this could be consolidated by using the `TryGetValue` method with an `out var` to capture the value if it exists.

File: src/Core/ExecutionPathTracer/ExecutionPathTracer.cs [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+
+            // Get most recent measurement on given control qubit
+            var cId = this.classicalRegisters[qId].Count - 1;
+            return this.classicalRegisters[qId][cId];
+        }
+
+        /// <summary>
+        /// Given a <c>Type</c> and its value, extract its fields and format it as a string.
+        /// </summary>
+        private string? ExtractArgs(Type t, object value)
```
Comment (cgranade) [inline src/Core/ExecutionPathTracer/ExecutionPathTracer.cs:164]: Does this need to be an instance method, or can it be a static extension method?

File: src/Core/ExecutionPathTracer/ExecutionPathTracer.cs [hunk 8]
```diff
@@ -0,0 +1,307 @@
...
+        {
+            List<string?> fields = new List<string?>();
```
Comment (cgranade) [inline src/Core/ExecutionPathTracer/ExecutionPathTracer.cs:166]:
```suggestion
            var fields = new List<string?>();
```

File: src/Core/ExecutionPathTracer/ExecutionPathTracer.cs [hunk 9]
```diff
@@ -0,0 +1,307 @@
...
+            foreach (var f in t.GetFields())
+            {
+                // If field is a tuple, recursively extract its inner arguments and format as a tuple string.
+                if (f.FieldType.IsTuple())
+                {
+                    var nestedArgs = f.GetValue(value);
+                    if (nestedArgs != null)
+                    {
+                        var nestedFields = this.ExtractArgs(f.FieldType, nestedArgs);
+                        fields.Add(nestedFields);
```
Comment (cgranade) [inline src/Core/ExecutionPathTracer/ExecutionPathTracer.cs:177]: Having a `.Add` method in a foreach loop is a great code smell for an opportunity to use `Select` instead. The trick here will be that you may want SelectMany to flatten out tuples...

File: src/Core/ExecutionPathTracer/ExecutionPathTracer.cs (hunk 7 above)
Comment (cgranade) [inline src/Core/ExecutionPathTracer/ExecutionPathTracer.cs:164]: The name `Extract` doesn't communicate that this works to format as a string as well.

Comment (cgranade) [inline src/Core/ExecutionPathTracer/ExecutionPathTracer.cs:164]: When should this method return null?

File: src/Core/ExecutionPathTracer/Register.cs [hunk 10]
```diff
@@ -0,0 +1,77 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (cgranade) [inline src/Core/ExecutionPathTracer/Register.cs:1]:
```suggestion
// Copyright (c) Microsoft Corporation.
```

File: src/Core/ExecutionPathTracer/Register.cs [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+// Licensed under the MIT License.
+
```
Comment (cgranade) [inline src/Core/ExecutionPathTracer/Register.cs:3]:
```suggestion
#nullable enable

```

File: src/Core/ExecutionPathTracer/Register.cs [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+namespace Microsoft.Quantum.IQSharp
+{
+    /// <summary>
+    /// Enum for the 2 types of registers: Qubit and Classical.
+    /// </summary>
+    public enum RegisterType
+    {
+        Qubit,
+        Classical,
```
Comment (cgranade) [inline src/Core/ExecutionPathTracer/Register.cs:12]: Since this is a public enum, some API documentation comments would be really helpful here.

File: src/Core/Core.csproj [hunk 13]
```diff
@@ -34,11 +34,13 @@
     <PackageReference Include="Microsoft.CSharp" Version="4.5.0" />
     <PackageReference Include="Microsoft.Extensions.DependencyInjection.Abstractions" Version="3.0.0" />
     <PackageReference Include="Microsoft.Extensions.Logging" Version="3.0.0" />
+    <PackageReference Include="Microsoft.Quantum.Canon" Version="0.5.1904.1302" />
```
Comment (rmshaffer) [inline src/Core/Core.csproj:37]: Just curious, what are you using from this package? I don't think `Microsoft.Quantum.Canon` is what you want, as that one has been deprecated for quite some time.

Comment (cgranade) [reply src/Core/Core.csproj:37]: If you're looking for `typeof(Microsoft.Quantum.Canon.ApplyToEach)`, you'll want Microsoft.Quantum.Standard here instead. That said, the libraries repo depends on IQ# such that that then creates a circular dependency between repositories.

File: src/Core/ExecutionPathTracer/ExecutionPath.cs [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+        public ExecutionPath(QubitDeclaration[] qubits, Operation[] operations)
+        {
+            this.Qubits = qubits;
+            this.Operations = operations;
+        }
+
+        /// <summary>
+        /// An array of <c>QubitDeclaration</c>s that represents the declared qubits used in the execution path.
+        /// </summary>
+        public QubitDeclaration[] Qubits { get; set; }
```
Comment (rmshaffer) [inline src/Core/ExecutionPathTracer/ExecutionPath.cs:32]: Can these properties use `private set;` to prevent consumers from changing them after object creation?

File: src/Core/ExecutionPathTracer/ExecutionPath.cs [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+            this.NumChildren = numChildren;
+        }
+
+        /// <summary>
+        /// Id of qubit.
+        /// </summary>
+        public int Id { get; set; }
+
+        /// <summary>
+        /// Number of associated classical registers.
```
[the remark this answers is not in the document]
Comment (rmshaffer) [reply src/Core/ExecutionPathTracer/ExecutionPath.cs:83]: Same question... if you can just use `0` as a default instead of `null`, it may be simpler, unless `0` and `null` have different meanings here.

File: src/Core/ExecutionPathTracer/ExecutionPath.cs [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+        /// </summary>
+        public string ArgStr { get; set; }
+
+        /// <summary>
+        /// Group of operations for each classical branch.
+        /// </summary>
+        /// <remarks>
+        /// Currently not used as this is intended for classically-controlled operations.
+        /// </remarks>
+        public List<List<Operation>> Children { get; set; }
```
Comment (rmshaffer) [inline src/Core/ExecutionPathTracer/ExecutionPath.cs:109]: Can this use `IEnumerable` rather than `List`? Also, similar comment as elsewhere about using `private set` wherever possible.

File: src/Core/ExecutionPathTracer/ExecutionPath.cs [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+
+        /// <summary>
+        /// List of control registers.
+        /// </summary>
+        public List<Register> Controls { get; set; } = new List<Register>();
+
+        /// <summary>
+        /// List of target registers.
+        /// </summary>
+        public List<Register> Targets { get; set; } = new List<Register>();
```
Comment (rmshaffer) [inline src/Core/ExecutionPathTracer/ExecutionPath.cs:129]: Same here, consider using `IEnumerable` instead of `List` for the property type. Of course, using `new List` is still totally fine. 😊

File: src/Core/ExecutionPathTracer/ExecutionPath.cs [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ExecutionPathTracer"/> class.
+        /// </summary>
+        /// <param name="qubits">
+        /// An array of <c>QubitDeclaration</c>s that represents the declared qubits used in the execution path.
+        /// </param>
+        /// <param name="operations">
+        /// An array of <c>Operation</c>s that represents the operations used in the execution path.
+        /// </param>
+        public ExecutionPath(QubitDeclaration[] qubits, Operation[] operations)
```
[the remark this answers is not in the document]
Comment (theRoughCode) [reply src/Core/ExecutionPathTracer/ExecutionPath.cs:23]: Yea, they can be `IEnumerable`s as long as the JSON serializer converts it into an array (which should be the case).

File: src/Core/ExecutionPathTracer/ExecutionPath.cs [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+    /// <summary>
+    /// Represents the qubit resources and operations traced out in an execution path of a Q# operation.
+    /// </summary>
+    public class ExecutionPath
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ExecutionPathTracer"/> class.
+        /// </summary>
+        /// <param name="qubits">
+        /// An array of <c>QubitDeclaration</c>s that represents the declared qubits used in the execution path.
```
[the remark this answers is not in the document]
Comment (theRoughCode) [reply src/Core/ExecutionPathTracer/ExecutionPath.cs:18]: Will do!

File: src/Core/ExecutionPathTracer/ExecutionPathTracer.cs (hunk 7 above)
Comment (theRoughCode) [reply src/Core/ExecutionPathTracer/ExecutionPathTracer.cs:164]: Ah, I guess that could've been an artifact of what it was doing initially (extracting it to a list of string). I guess `ArgsToString` would be a better name.

Comment (theRoughCode) [reply src/Core/ExecutionPathTracer/ExecutionPathTracer.cs:164]: It returns null if there aren't any arguments found. This was so that the `argStr` value in the resulting `Operation` object would be null and wouldn't be rendered in the final JSON string.

File: src/Core/Core.csproj (hunk 13 above)
Comment (theRoughCode) [reply src/Core/Core.csproj:37]: Yea, @cgranade is right, it was to resolve `Microsoft.Quantum.Canon.ApplyToEach`. I did import `Microsoft.Quantum.Standard` in the first commit, but that broke some tests. Any ideas on why that broke it vs using `Microsoft.Quantum.Canon` would be appreciated!

I guess moving it to `qsharp-runtime` would help solve this. cc-ing @anpaz-msft regarding thoughts on this.

File: src/Core/ExecutionPathTracer/ExecutionPath.cs (hunk 14 above)
Comment (theRoughCode) [reply src/Core/ExecutionPathTracer/ExecutionPath.cs:32]: Yea, these objects aren't modified after creation, so we can definitely do that.

File: src/Core/ExecutionPathTracer/ExecutionPath.cs (hunk 15 above)
[the remark this answers is not in the document]
Comment (theRoughCode) [reply src/Core/ExecutionPathTracer/ExecutionPath.cs:83]: The `null` value here is so that this key doesn't get rendered in the JSON string. What are your thoughts on making it `null` so that it doesn't get rendered vs making it `0` for simplicity?

File: src/Core/ExecutionPathTracer/ExecutionPath.cs [hunk 20]
```diff
@@ -0,0 +1,131 @@
...
+    {
+        /// <summary>
+        /// Label of gate.
+        /// </summary>
+        public string Gate { get; set; }
+
+        /// <summary>
+        /// Non-qubit arguments provided to gate.
+        /// </summary>
+        public string ArgStr { get; set; }
```
[the remark this answers is not in the document]
Comment (theRoughCode) [reply src/Core/ExecutionPathTracer/ExecutionPath.cs:101]: We only want to render the non-qubit arguments in the argument string when we visualize them as the qubits they act on are already implied by the visualization. Would `Arguments` be a better property name?

File: src/Core/ExecutionPathTracer/ExecutionPathTracer.cs (hunk 1 above)
Comment (theRoughCode) [reply src/Core/ExecutionPathTracer/ExecutionPathTracer.cs:56]: That's cool! I guess we could use it with something like this:
```C#
qubitDecl = qubitDecl with {
  NumChildren = (this.classicalRegisters.ContainsKey(k))
    ? this.classicalRegisters[k].Count
    : null
};
```

Comment (cgranade) [reply src/Core/ExecutionPathTracer/ExecutionPathTracer.cs:56]: Yep! Doesn't work for us yet since this project is still on C# 8, but it's exciting to keep an eye on where new features can help in the future.

File: src/Core/ExecutionPathTracer/ExecutionPathTracer.cs (hunk 2 above)
Comment (theRoughCode) [reply src/Core/ExecutionPathTracer/ExecutionPathTracer.cs:104]: Would `GetOrCreate*` be better?

File: src/Core/ExecutionPathTracer/ExecutionPathTracer.cs (hunk 4 above)
Comment (theRoughCode) [reply src/Core/ExecutionPathTracer/ExecutionPathTracer.cs:124]: Oh, interesting, that would make sense. Thanks!

File: src/Core/ExecutionPathTracer/ExecutionPathTracer.cs (hunk 6 above)
Comment (theRoughCode) [reply src/Core/ExecutionPathTracer/ExecutionPathTracer.cs:108]: Do you mean something similar to [this](https://stackoverflow.com/questions/16192906/net-dictionary-get-or-create-new)? Not sure what you mean by using an `out var` to capture the value.

File: src/Core/ExecutionPathTracer/ExecutionPathTracer.cs (hunk 7 above)
Comment (theRoughCode) [reply src/Core/ExecutionPathTracer/ExecutionPathTracer.cs:164]: This can definitely be a static method. Thanks!

File: src/Core/ExecutionPathTracer/ExecutionPathTracer.cs (hunk 9 above)
Comment (theRoughCode) [reply src/Core/ExecutionPathTracer/ExecutionPathTracer.cs:177]: Ah, thanks for the tip! Still getting used to the tricks of C# :)

File: src/Core/ExecutionPathTracer/ExecutionPath.cs (hunk 17 above)
Comment (theRoughCode) [reply src/Core/ExecutionPathTracer/ExecutionPath.cs:129]: Thanks for the tip! Really appreciate it :D

File: src/Core/ExecutionPathTracer/ExecutionPath.cs (hunk 16 above)
Comment (theRoughCode) [reply src/Core/ExecutionPathTracer/ExecutionPath.cs:109]: I think in this case, since we don't have a constructor, we probably can't use `private set` here? Not sure what the convention is, but there would be a lot of arguments to pass into the construction for the `Operation` class, so I figured manually setting it would be the best way.

File: src/Core/ExecutionPathTracer/ExecutionPathTracer.cs (hunk 6 above)
Comment (theRoughCode) [reply src/Core/ExecutionPathTracer/ExecutionPathTracer.cs:108]: I created an extension method like [this](https://stackoverflow.com/a/16193323) for reusability.

File: src/Core/Core.csproj (hunk 13 above)
Comment (cgranade) [reply src/Core/Core.csproj:37]: @theRoughCode: The Q# libraries are my area, so I'm happy to help with and to discuss. The short version is that we have sometimes moved operations from the Q# standard libraries into the Q# core library to make them available to the simulation runtime and to IQ# (e.g.: `MResetZ`). This doesn't have any user-facing effect, as both the core and standard libraries are made available to Q# applications, but can help with dependency ordering. From that perspective, moving ApplyToEach* to the core so that they can be used here is a completely reasonable suggestion.

The flip side is that it's always a bit of an awkward cross-repo dance to move something from standard to core, so I'd want to be quite sure that's the best solution before proceeding.

File: src/Core/ExecutionPathTracer/ExecutionPathTracer.cs (hunk 7 above)
Comment (cgranade) [reply src/Core/ExecutionPathTracer/ExecutionPathTracer.cs]: That makes sense; I'd suggest including some version of that discussion in the documentation comment, perhaps as a `<returns>` element.

File: src/Core/ExecutionPathTracer/ExecutionPath.cs (hunk 20 above)
[the remark this answers is not in the document]
Comment (cgranade) [reply src/Core/ExecutionPathTracer/ExecutionPath.cs]: I'm a bit mixed here, as there's some potential for confusion with Q# syntax. That is, qubits are just ordinary values of type `Qubit`, such that the input tuple would look confusing if you just pulled out the qubits. For example, how would this property represent an input of type [`FixedPoint`](https://docs.microsoft.com/en-us/qsharp/api/qsharp/microsoft.quantum.arithmetic.fixedpoint), since that UDT is declared as having `(Int, Qubit[])` as its base type?

File: src/Core/Core.csproj (hunk 13 above)
Comment (theRoughCode) [reply src/Core/Core.csproj:37]: @cgranade Yea, that makes sense. I was talking to Andres on a sync earlier and he suggested I restructure the core logic surrounding that parsing of each operation into the metadata `Operation` (i.e. `ParseOperation` function) into qsharp-runtime. I'll be looping you in as I make progress on that front. This PR might be reworked a bit till then, so I'll make this a draft PR for now to signify that.

File: src/Core/ExecutionPathTracer/ExecutionPath.cs (hunk 20 above)
[the remark this answers is not in the document]
Comment (theRoughCode) [reply src/Core/ExecutionPathTracer/ExecutionPath.cs]: That would most likely (if represented as a tuple type) would return `(Int)`. So, if we have `Foo(FixedPoint)` the visualization would create a gate acting on `Qubit[]` wires and the label `Foo(Int)`.

File: src/Core/ExecutionPathTracer/ExecutionPath.cs (hunk 15 above)
[the remark this answers is not in the document]
Comment (rmshaffer) [reply src/Core/ExecutionPathTracer/ExecutionPath.cs:84]: Is there any case where `0` would mean something different than `null`? My preference would be to just use `0` and make the property non-nullable, as `NumChildren == null` doesn't seem very intuitive. You can just ignore this if it's `0` when parsing the JSON, correct?

File: src/Core/ExecutionPathTracer/ExecutionPathTracer.cs (hunk 2 above)
Comment (rmshaffer) [reply src/Core/ExecutionPathTracer/ExecutionPathTracer.cs:102]: In general I agree with @cgranade's comment, but in this particular case I like `Get*`, since it's basically just allowing a lazy-initialization of the `qubitRegisters` dictionary. Up to you, though, no big deal either way.

File: src/Core/ExecutionPathTracer/ExecutionPath.cs (hunk 15 above)
[the remark this answers is not in the document]
Comment (theRoughCode) [reply src/Core/ExecutionPathTracer/ExecutionPath.cs:84]: Yea, I agree. I think I was looking into this for `System.Text.Json` which I'm using, but there isn't a nice way _yet_ to ignore a property based on a value. I _think_ this is possible with `Newtonsoft`?

File: src/Core/ExecutionPathTracer/ExecutionPathTracer.cs (hunk 2 above)
Comment (theRoughCode) [reply src/Core/ExecutionPathTracer/ExecutionPathTracer.cs:81]: Cool, thanks for the input! I'll stick to `Get*` unless there are strong objections against this. :)

File: src/Core/ExecutionPathTracer/ExecutionPath.cs (hunk 20 above)
[the remark this answers is not in the document]
Comment (theRoughCode) [reply src/Core/ExecutionPathTracer/ExecutionPath.cs]: Changed the name to `DisplayArgs` to denote that these might not contain all arguments, only the arguments we want to display.

File: src/Core/ExecutionPathTracer/ExecutionPathTracer.cs [hunk 21]
```diff
@@ -0,0 +1,150 @@
...
+                Controls = this.GetQubitRegisters(metadata.Controls),
+                Targets = this.GetQubitRegisters(metadata.Targets),
+            };
+
+            // Create classical registers for measurement operations
+            if (metadata.IsMeasurement)
+            {
+                var measureQubit = metadata.Targets.ElementAt(0);
+                var clsReg = this.CreateClassicalRegister(measureQubit);
+                // TODO: Change this to using IsMeasurement
```
Comment (theRoughCode) [inline src/Core/ExecutionPathTracer/ExecutionPathTracer.cs:141]: Note: this will be addressed in a separate PR so as not to clutter this one.
